File: enchiridion-go/internal/ingestignore/ingestignore_test.go
package ingestignore

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestParse(t *testing.T) {
	patterns, err := Parse("# a comment\n\nliteral.md\n*.tmp  # trailing comment\n   \n")
	if err != nil {
		t.Fatalf("Parse: %v", err)
	}
	if strings.Join(patterns, ",") != "literal.md,*.tmp" {
		t.Errorf("Parse = %v, want [literal.md *.tmp]", patterns)
	}
}

func TestParseRejectsRicherPatterns(t *testing.T) {
	for _, line := range []string{"sub/dir.md", "!keep.md", "**/deep.md"} {
		if _, err := Parse(line); err == nil {
			t.Errorf("Parse(%q): want an error, got nil", line)
		}
	}
}

func TestAppendCreatesFile(t *testing.T) {
	folder := t.TempDir()
	if err := Append(folder, "doc.md", "ingested before back-pointers were mandatory"); err != nil {
		t.Fatalf("Append: %v", err)
	}
	text := readIgnore(t, folder)
	want := "doc.md  # ingested before back-pointers were mandatory\n"
	if text != want {
		t.Errorf("file = %q, want %q", text, want)
	}
}

func TestAppendIsIdempotent(t *testing.T) {
	folder := t.TempDir()
	for range 3 {
		if err := Append(folder, "doc.md", ""); err != nil {
			t.Fatalf("Append: %v", err)
		}
	}
	if text := readIgnore(t, folder); text != "doc.md\n" {
		t.Errorf("file = %q, want a single entry", text)
	}
}

func TestAppendPreservesExistingContent(t *testing.T) {
	folder := t.TempDir()
	if err := os.WriteFile(filepath.Join(folder, Filename), []byte("# policy\n*.tmp\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	if err := Append(folder, "doc.md", ""); err != nil {
		t.Fatalf("Append: %v", err)
	}
	if text := readIgnore(t, folder); text != "# policy\n*.tmp\ndoc.md\n" {
		t.Errorf("file = %q", text)
	}
}

// The folder always exists in a real call — the file being withdrawn is
// sitting in it. A missing one means a mistyped path, and must not be
// papered over by creating it.
func TestAppendRefusesMissingFolder(t *testing.T) {
	folder := filepath.Join(t.TempDir(), "emials")
	if err := Append(folder, "doc.eml", ""); err == nil {
		t.Error("Append into a folder that doesn't exist: want an error, got nil")
	}
	if _, err := os.Stat(folder); err == nil {
		t.Error("Append created the mistyped folder")
	}
}

func readIgnore(t *testing.T, folder string) string {
	t.Helper()
	text, err := os.ReadFile(filepath.Join(folder, Filename))
	if err != nil {
		t.Fatal(err)
	}
	return string(text)
}

File: enchiridion-ts/src/ingestignore.test.ts
/**
 * ingestignore tests — parse and append to the per-folder .ingestignore.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { Filename, append, parse } from "./ingestignore.js";

test("parse strips comments and blanks", () => {
  const patterns = parse(
    "# a comment\n\nliteral.md\n*.tmp  # trailing comment\n   \n",
  );
  assert.deepEqual(patterns, ["literal.md", "*.tmp"]);
});

test("parse rejects richer patterns", () => {
  for (const line of ["sub/dir.md", "!keep.md", "**/deep.md"]) {
    assert.throws(() => parse(line));
  }
});

test("append creates the file", () => {
  const folder = fs.mkdtempSync(path.join(os.tmpdir(), "enchiridion-ii-"));
  append(folder, "doc.md", "ingested before back-pointers were mandatory");
  const text = readIgnore(folder);
  assert.equal(
    text,
    "doc.md  # ingested before back-pointers were mandatory\n",
  );
});

test("append is idempotent", () => {
  const folder = fs.mkdtempSync(path.join(os.tmpdir(), "enchiridion-ii-"));
  for (let i = 0; i < 3; i++) append(folder, "doc.md", "");
  assert.equal(readIgnore(folder), "doc.md\n");
});

test("append preserves existing content", () => {
  const folder = fs.mkdtempSync(path.join(os.tmpdir(), "enchiridion-ii-"));
  fs.writeFileSync(path.join(folder, Filename), "# policy\n*.tmp\n");
  append(folder, "doc.md", "");
  assert.equal(readIgnore(folder), "# policy\n*.tmp\ndoc.md\n");
});

test("append refuses a missing folder", () => {
  const folder = path.join(
    fs.mkdtempSync(path.join(os.tmpdir(), "enchiridion-ii-")),
    "emials",
  );
  assert.throws(() => append(folder, "doc.eml", ""));
  assert.equal(fs.existsSync(folder), false);
});

function readIgnore(folder: string): string {
  return fs.readFileSync(path.join(folder, Filename), "utf8");
}
